hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/supervisor/workload/SelectiveSupervisorControllersIntegrationTest.java

```diff
@@ -0,0 +1,224 @@
+
+        // then
+        waitForRegistration(id);
+    }
+
+    @Test
+    public void shouldElectOnlyOneLeaderFromRegisteredConsumers() {
+        // given
+        List<SelectiveSupervisorController> supervisors = ImmutableList.of(
+                getConsumerSupervisor("1"), getConsumerSupervisor("2"), getConsumerSupervisor("3"));
+
+        // when
+        supervisors.forEach(this::startConsumer);
+
+        // then
+        assertThat(supervisors.stream().filter(SelectiveSupervisorController::isLeader).count()).isEqualTo(1);
+    }
+
+    @Test
+    public void shouldElectNewLeaderAfterShutdown() throws InterruptedException {
+        // given
+        Map<String, CuratorFramework> curators = ImmutableMap.of("A", otherClient(), "B", otherClient());
+        List<SelectiveSupervisorController> supervisors = curators.entrySet().stream()
+                .map(entry -> getConsumerSupervisor(entry.getKey(), entry.getValue()))
+                .collect(Collectors.toList());
+        supervisors.forEach(this::startConsumer);
+        SelectiveSupervisorController leader = findLeader(supervisors);
+
+        // when
+        curators.get(leader.getId()).close();
+        await().atMost(adjust(ONE_SECOND)).until(() -> !consumersRegistry.isRegistered(leader.getId()));
+
+        // then
+        await().atMost(adjust(TEN_SECONDS)).until(() -> supervisors.stream()
+                .filter(SelectiveSupervisorController::isLeader)
+                .filter(c -> !c.equals(leader))
+                .findAny().isPresent());
+
+        await().atMost(adjust(ONE_SECOND)).until(() -> !leader.isLeader());
+    }
+
+    @Test
+    public void shouldAssignConsumerToSubscription() {
+        // given
+        SelectiveSupervisorController controller = getConsumerSupervisor("c1");
+        startConsumer(controller);
+
+        // when
+        createSubscription("com.example.topic$test");
+
+        // then
+        awaitUntilAssignmentExists("com.example.topic$test", "c1");
+    }
+
+    @Test
+    public void shouldAssignSubscriptionToMultipleConsumers() {
+        // given
+        ImmutableList.of(getConsumerSupervisor("c1"), getConsumerSupervisor("c2")).forEach(this::startConsumer);
+
+        // when
+        createSubscription("com.example.topic$test");
+
+        // then
+        awaitUntilAssignmentExists("com.example.topic$test", "c1");
+        awaitUntilAssignmentExists("com.example.topic$test", "c2");
+    }
+
+    @Test
+    public void shouldAssignConsumerToMultipleSubscriptions() {
+        // given
+        startConsumer(getConsumerSupervisor("c1"));
+
+        // when
+        createSubscription("com.example.topic$test1");
+        createSubscription("com.example.topic$test2");
+
+        // then
+        awaitUntilAssignmentExists("com.example.topic$test1", "c1");
+        awaitUntilAssignmentExists("com.example.topic$test2", "c1");
+    }
+
+    private void awaitUntilAssignmentExists(String subscription, String supervisorId) {
```

Comment:
those private methods could be moved to some fixture builder, as in previosu comment